### PR TITLE
refactor: wildcard resolution and pl::Expr::target_ids

### PR DESCRIPTION
**Pull request by aljazerzen**
This PR has partially been cherry-picked from an old branch of mine and partially rewritten anew. I've worked on this for a few weeks now, so I forgot the initial motivation of why I wanted to change this :D

It does simplify:
- the wildcard resolution (this opinion might be biased since I spent so much time debugging it and I know how it works in depth now)
- `Expr::target_id` and `Expr::target_ids`, which are now a single number.

I *think* I wanted this because it made lineage inference independent of the declarations in root module: https://github.com/PRQL/prql/pull/3863/files#diff-d2aed3357385f80cd4517d8ca5492ad503c33c13b8afbd9e219666f57f947459L264

As an additional upside, I changed functions implemented specifically for lineage into functions that manipulate the AST, which means that I don't have to duplicate this behavior in type inference.

### CODE REVIEW

File: justfile [hunk 1]
```diff
@@ -30,13 +30,13 @@ prqlc-lint:
 # Test prqlc
 packages := '--package=prqlc-ast --package=prqlc-parser --package=prql-compiler --package=prqlc'
 prqlc-test-fast:
-    cargo clippy --all-targets {{packages}} -- -D warnings
-
     # cargo insta test, but allowing multiple --package arguments
     INSTA_FORCE_PASS=1 cargo test --locked {{packages}}
```
Comment (max-sixty) [inline justfile:34]: FYI this is now fixed upstream in `cargo-insta` so you can use it directly with multiple packages!

Comment (aljazerzen) [reply justfile:34]: Just in time for me to discover cargo-nextest :D

Comment (max-sixty) [reply justfile:34]: Also supported by `cargo-insta`! :D

**Comment by max-sixty** [comment]
Some of the changes in test outputs look good!

**Comment by aljazerzen** [comment]
I realized now that I didn't even mention why I'm changing this! I've edited the original message.

**Pull request merged**